definitions for linux hardware timestamping

folkertdev opened the pull request:
Definitions can be found here https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/net_tstamp.h#L76

These definitions are relevant for (PTP and NTP) hardware timestamping

folkertdev commented:
I'm not sure what this CI error means

```
  cargo:warning=/checkout/target/i686-unknown-linux-gnu/debug/build/libc-test-1d150dc772ee3115/out/main.c:35676:80: error: 'HWTSTAMP_FLAG_BONDED_PHC_INDEX' undeclared here (not in a function)
  cargo:warning=35676 |             static const int __test_const_HWTSTAMP_FLAG_BONDED_PHC_INDEX_val = HWTSTAMP_FLAG_BONDED_PHC_INDEX;
  cargo:warning=      |                                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=/checkout/target/i686-unknown-linux-gnu/debug/build/libc-test-1d150dc772ee3115/out/main.c:35682:68: error: 'HWTSTAMP_FLAG_LAST' undeclared here (not in a function)
  cargo:warning=35682 |             static const int __test_const_HWTSTAMP_FLAG_LAST_val = HWTSTAMP_FLAG_LAST;
  cargo:warning=      |                                                                    ^~~~~~~~~~~~~~~~~~
  cargo:warning=/checkout/target/i686-unknown-linux-gnu/debug/build/libc-test-1d150dc772ee3115/out/main.c:35688:68: error: 'HWTSTAMP_FLAG_MASK' undeclared here (not in a function)
  cargo:warning=35688 |             static const int __test_const_HWTSTAMP_FLAG_MASK_val = HWTSTAMP_FLAG_MASK;
  cargo:warning=      |                                                                    ^~~~~~~~~~~~~~~~~~
  cargo:warning=cc1: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
```

maybe the system libc on CI does not (yet?) have these symbols? They are not important to me so I'd be OK with removing them (just thought I might as well add them for completeness)

JohnTitor commented:
> maybe the system libc on CI does not (yet?) have these symbols?

Correct.

Could you squash commits into one?

JohnTitor commented:
Thank you, @bors r+

bors commented:
:pushpin: Commit b409bdca8f42faa11e26cac397669ca6361ce220 has been approved by `JohnTitor`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/libc) for this repository.

<!-- @bors r=JohnTitor b409bdca8f42faa11e26cac397669ca6361ce220 -->
<!-- homu: {"type":"Approved","sha":"b409bdca8f42faa11e26cac397669ca6361ce220","approver":"JohnTitor","queue":"https://bors.rust-lang.org/homu/queue/libc"} -->

bors commented:
:hourglass: Testing commit b409bdca8f42faa11e26cac397669ca6361ce220 with merge 38ff53af6b4a9df70714621917a229e4fb00fa4a...
<!-- homu: {"type":"BuildStarted","head_sha":"b409bdca8f42faa11e26cac397669ca6361ce220","merge_sha":"38ff53af6b4a9df70714621917a229e4fb00fa4a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/libc/actions/runs/4499846464/jobs/7918454645)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/libc/actions/runs/4499846464/jobs/7918454645","builder_name":"checks-actions"} -->

folkertdev commented:
I'm not sure what went wrong here. I assume the action needs to just be re-run?

folkertdev commented:
I added two extra constants that are needed (`SIOCSHWTSTAMP` and `SIOCGHWTSTAMP`) and rebased. Maybe that makes bors happier?

though CI seems to be broken at the moment.

JohnTitor commented:
The FreeBSD failure is unrelated and it's already fixed on master.
But https://github.com/rust-lang/libc/pull/3155#issuecomment-1481006102 is legit. That's because (probably) musl toolchain on CI is old, could you skip tests on that env?

folkertdev commented:
ok, rebased again to fix that freebsd problem

> could you skip tests on that env

how would I do this? do you mean a `cfg` for the undefined constants?

JohnTitor commented:
https://github.com/rust-lang/libc/blob/master/libc-test/build.rs is a place to tweak, like:
https://github.com/rust-lang/libc/blob/68e06adafa8f793b094a23b210d39b27c813600c/libc-test/build.rs#L1739-L1754

folkertdev commented:
allright, I added logic to ignore the constant for `i686` and `mips`. that should make the tests that failed before pass, but there might be other (cancelled) ci runs that also should skip this constant. Is there a better way to figure that out than trial and error?

JohnTitor commented:
There's a trade-off, we could disable the fail-fast but we may want to find failures fast.
@bors r+

bors commented:
:pushpin: Commit 5842b24052a9fe350b6f4437615d28bb16c9f3b6 has been approved by `JohnTitor`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/libc) for this repository.

<!-- @bors r=JohnTitor 5842b24052a9fe350b6f4437615d28bb16c9f3b6 -->
<!-- homu: {"type":"Approved","sha":"5842b24052a9fe350b6f4437615d28bb16c9f3b6","approver":"JohnTitor","queue":"https://bors.rust-lang.org/homu/queue/libc"} -->

JohnTitor commented:
(Another way would be running the same command as CI on your local.)

JohnTitor commented on libc-test/build.rs:
```diff
@@ -3724,6 +3724,8 @@ fn test_linux(target: &str) {
                 => true,
             "SCTP_FUTURE_ASSOC" | "SCTP_CURRENT_ASSOC" | "SCTP_ALL_ASSOC" | "SCTP_PEER_ADDR_THLDS_V2" => true, // linux 5.5+
 
+            "HWTSTAMP_TX_ONESTEP_P2P" if i686 || mips || aarch64_musl => true,
```
I think the below should work like others:

```suggestion
            // FIXME: Requires more recent kernel headers
            "HWTSTAMP_TX_ONESTEP_P2P" if mips || musl => true,
```

folkertdev replied:
yes, fixed. A couple more linux flavors to get through

JohnTitor replied:
The condition doesn't seem to be changed. It should `mips || musl` as we use older kernels for these targets.

folkertdev replied:
though maybe musl is more relevant than aarch here?

JohnTitor replied:
We use the same kernel version for _all_ the musl targets across CI: https://github.com/rust-lang/libc/blob/c64bfc5fb36414e02ad7977b32df765876632383/ci/install-musl.sh#L82
So, CPU arch doesn't involve here, we just need to ignore `musl` on the if cond (the same applies to sparc64).

folkertdev replied:
so `i686 || mips || musl`, or only `musl` even?

JohnTitor replied:
Ah, I mean the condition should be `sparc64 || musl`.

JohnTitor replied:
For sparc64, see:
https://github.com/rust-lang/libc/blob/c64bfc5fb36414e02ad7977b32df765876632383/ci/docker/sparc64-unknown-linux-gnu/Dockerfile#L1-L2

JohnTitor replied:
i686 itself isn't related to the failure, it fails due to the musl toolchain.

folkertdev replied:
makes sense, I've updated it

JohnTitor commented:
Let's try again, @bors r+

bors commented:
:pushpin: Commit 58ffdf4226dcba271d55bb0868dc4b3e80f34f68 has been approved by `JohnTitor`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/libc) for this repository.

<!-- @bors r=JohnTitor 58ffdf4226dcba271d55bb0868dc4b3e80f34f68 -->
<!-- homu: {"type":"Approved","sha":"58ffdf4226dcba271d55bb0868dc4b3e80f34f68","approver":"JohnTitor","queue":"https://bors.rust-lang.org/homu/queue/libc"} -->

bors commented:
:hourglass: Testing commit 58ffdf4226dcba271d55bb0868dc4b3e80f34f68 with merge c8de3b4856e97e19dc801c58f9735dfc57e0e1cf...
<!-- homu: {"type":"BuildStarted","head_sha":"58ffdf4226dcba271d55bb0868dc4b3e80f34f68","merge_sha":"c8de3b4856e97e19dc801c58f9735dfc57e0e1cf"} -->

folkertdev commented:
we're getting closer. I'm not sure what this is trying to do though:

```
RUNNING ALL TESTS
rust[8] = 0 != 8 (C): Rust "hwtstamp_config" -> C
rust[9] = 0 != 9 (C): Rust "hwtstamp_config" -> C
rust[10] = 0 != 10 (C): Rust "hwtstamp_config" -> C
rust[11] = 0 != 11 (C): Rust "hwtstamp_config" -> C
thread 'main' panicked at 'some tests failed', /checkout/target/sparc64-unknown-linux-gnu/debug/build/libc-test-298fd54f1fef40cd/out/main.rs:12:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `--test main`

Caused by:
  process didn't exit successfully: `/test-runner-linux sparc64 /checkout/target/sparc64-unknown-linux-gnu/debug/deps/main-973f2bcc439c5f8d` (exit status: 1)
```

should we skip the `hwtstamp_config` struct on this target?

folkertdev commented:
rather looks like this is an ABI roundtrip test. That this fails is extremely surprising to me, given that the struct is just three `int`s, so nothing tricky is going on. Should we just ignore the roundtrip for `sparc64`?

JohnTitor commented:
I guess https://github.com/rust-lang/libc/issues/1558 is related. Yeah, skipping would be fine 👍

folkertdev commented:
allright let's hope that's it

JohnTitor commented:
@bors r+

bors commented:
:pushpin: Commit 494999832f7342ab3c3ef8123a45fc69f434f1ab has been approved by `JohnTitor`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/libc) for this repository.

<!-- @bors r=JohnTitor 494999832f7342ab3c3ef8123a45fc69f434f1ab -->
<!-- homu: {"type":"Approved","sha":"494999832f7342ab3c3ef8123a45fc69f434f1ab","approver":"JohnTitor","queue":"https://bors.rust-lang.org/homu/queue/libc"} -->

bors commented:
:hourglass: Testing commit 494999832f7342ab3c3ef8123a45fc69f434f1ab with merge 2dc31a7f72b64d04346f0f1f4bb7e57cafd1ad5d...
<!-- homu: {"type":"BuildStarted","head_sha":"494999832f7342ab3c3ef8123a45fc69f434f1ab","merge_sha":"2dc31a7f72b64d04346f0f1f4bb7e57cafd1ad5d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/libc/actions/runs/4601638660/jobs/8130556439), [checks-cirrus-freebsd-12](https://cirrus-ci.com/task/5783817793830912), [checks-cirrus-freebsd-13](https://cirrus-ci.com/task/5220867840409600), [checks-cirrus-freebsd-14](https://cirrus-ci.com/task/6346767747252224)
Approved by: JohnTitor
Pushing 2dc31a7f72b64d04346f0f1f4bb7e57cafd1ad5d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"JohnTitor","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/libc/actions/runs/4601638660/jobs/8130556439","checks-cirrus-freebsd-12":"https://cirrus-ci.com/task/5783817793830912","checks-cirrus-freebsd-13":"https://cirrus-ci.com/task/5220867840409600","checks-cirrus-freebsd-14":"https://cirrus-ci.com/task/6346767747252224"},"merge_sha":"2dc31a7f72b64d04346f0f1f4bb7e57cafd1ad5d"} -->
